# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: toolkit/components/credentialmanagement/IdentityCredentialRequestManager.h
# Commit: 63d9972bab1f
# Full Hash: 63d9972bab1f267a940ec30c7cea7a4c41e54fa6
# Author: Alexandru Marc <amarc@mozila.com>
# Date: 2025-06-06 08:52:03
# Regressor Bug: 1969017
# File Overlap Count: 2
# Description:
#   Revert "Bug 1969015, 1969017: apply code formatting via Lando" for causing leaks @ mochitest.toml
#   
#   This reverts commit a97cb1d6c08bc38e8644bfc916298b847dbac3a1.
#   
#   Revert "Bug 1969017 - Refactor FedCM to use a managed actor rather than tagging onto WindowGlobalParent - r=anti-tracking-reviewers,emz"
# ==============================================================================

diff -r 801e25cb4bfd -r 63d9972bab1f toolkit/components/credentialmanagement/IdentityCredentialRequestManager.h
--- a/toolkit/components/credentialmanagement/IdentityCredentialRequestManager.h	Thu Jun 05 18:51:36 2025 +0000
+++ b/toolkit/components/credentialmanagement/IdentityCredentialRequestManager.h	Thu Jun 05 21:40:49 2025 +0300
@@ -7,7 +7,13 @@
 #ifndef MOZILLA_IDENTITYCREDENTIALREQUESTMANAGER_H_
 #define MOZILLA_IDENTITYCREDENTIALREQUESTMANAGER_H_
 
-#include "nsISupports.h"
+#include "mozilla/ClearOnShutdown.h"
+#include "mozilla/PrincipalHashKey.h"
+#include "mozilla/dom/IdentityCredential.h"
+#include "mozilla/dom/IdentityCredentialBinding.h"
+#include "mozilla/dom/IPCIdentityCredential.h"
+#include "nsIPrincipal.h"
+#include "nsTHashMap.h"
 
 namespace mozilla {
 
@@ -17,6 +23,23 @@
 
   static IdentityCredentialRequestManager* GetInstance();
 
+  // Store an active cross origin identity credential request happening from the
+  // given principal and inner window ID. These accumulate forever, but if the
+  // window goes away, we will be unable to notify of a store.
+  nsresult StorePendingRequest(
+      const nsCOMPtr<nsIPrincipal>& aRPPrincipal,
+      const dom::IdentityCredentialRequestOptions& aRequest,
+      const RefPtr<
+          dom::IdentityCredential::GetIPCIdentityCredentialPromise::Private>&
+          aPromise,
+      const RefPtr<dom::CanonicalBrowsingContext>& aBrowsingContext);
+
+  // If the given credential stored by the given principal would be effective
+  // for a previously stored request, notify the window that stored that request
+  // with the credential so it can resolve a promise with that credential data.
+  void NotifyOfStoredCredential(const nsCOMPtr<nsIPrincipal>& aIDPPrincipal,
+                                const dom::IPCIdentityCredential& aCredential);
+
   IdentityCredentialRequestManager(IdentityCredentialRequestManager& other) =
       delete;
   void operator=(const IdentityCredentialRequestManager&) = delete;
@@ -25,6 +48,27 @@
   static StaticRefPtr<IdentityCredentialRequestManager> sSingleton;
   IdentityCredentialRequestManager() {};
   ~IdentityCredentialRequestManager() = default;
+
+  struct PendingRequestEntry {
+    nsCOMPtr<nsIPrincipal> mRPPrincipal;
+    dom::IdentityCredentialRequestOptions mRequestOptions;
+    RefPtr<dom::IdentityCredential::GetIPCIdentityCredentialPromise::Private>
+        mPromise;
+    RefPtr<dom::CanonicalBrowsingContext> mBrowsingContext;
+
+    PendingRequestEntry(
+        nsIPrincipal* aRPPrincipal,
+        const dom::IdentityCredentialRequestOptions& aRequestOptions,
+        const RefPtr<
+            dom::IdentityCredential::GetIPCIdentityCredentialPromise::Private>&
+            aPromise,
+        const RefPtr<dom::CanonicalBrowsingContext>& aBrowsingContext)
+        : mRPPrincipal(aRPPrincipal),
+          mRequestOptions(aRequestOptions),
+          mPromise(aPromise),
+          mBrowsingContext(aBrowsingContext) {}
+  };
+  nsTHashMap<PrincipalHashKey, nsTArray<PendingRequestEntry>> mPendingRequests;
 };
 
 }  // namespace mozilla